Add support for S3-compatible backends

tiwoc opened the pull request:
This PR adds support for backing up to AWS S3 buckets, and to other object storage services with an S3-compatible API such as Backblaze. In addition to an address similar to the one for SFTP and REST servers, S3 needs a key ID and a key. This PR adds fields for these, right below the address field, that only show up when S3 is selected:

<img width="624" src="https://github.com/user-attachments/assets/78d9dd37-3947-4bab-9a92-6eb0a74cf1a1" />

The values of these fields are then passed to Restic in the corresponding environment variables. The key ID is stored in UserDefaults, while the key is stored in the Keychain.

Restic docs:
- [Preparing a new repository: Amazon S3](https://restic.readthedocs.io/en/latest/030_preparing_a_new_repo.html#amazon-s3)
- [Preparing a new respository: S3-compatible storage](https://restic.readthedocs.io/en/latest/030_preparing_a_new_repo.html#s3-compatible-storage)

sergeymakinen commented:
Hi @tiwoc, thank you for the PR, this one looks good to me from the first sight!
I'm going to review this later after New Year, 2 things come into my mind right now:

- S3 key's fields should be grouped with the address field (and thus be padded from the password field below)
- I'd like to handle environment variables in some consistent way; not sure though but either constructing all the environment in the kit and just pass through the runner or keep doing what we're doing: pass S3 keys in distinct properties like we do with a password and keep assembling the environment in the runner

tiwoc commented:
Very good points, thanks for your feedback!

In the latest commits, I've now changed the field layout so that the S3 fields are grouped with the address field, and moved the creation of the env vars down to the runner for consistency. This probably doesn't scale to dozens of additional config options, but if there's no plan for adding many more, the current solution works well.

tiwoc commented:
Ah, I forgot to post an updated screenshot:
<img width="624" src="https://github.com/user-attachments/assets/5f2f1a5b-fc59-4d77-aa69-8d9d4dd0bb27" />

No pressure on the review though. I'm going to use my own build for the time being to see how it goes.

sergeymakinen commented on ResticScheduler/AppEnvironment.swift:
```diff
@@ -35,6 +35,8 @@ class AppEnvironment {
 
   @UserDefault("BackupFrequency") var backupFrequency = 86400
   @UserDefault("ResticRepository") var resticRepository = FileManager.default.temporaryDirectory.path(percentEncoded: false)
+  @UserDefault("ResticS3AccessKeyId") var s3AccessKeyId = ""
```
I think we should keep it optional (`String?`) like other optionals.

`s3SecretAccessKey` is optional too but `KeychainPassword` needs some updates from me to make it work, so I'll take care of it afterwards.

tiwoc replied:
Just to make sure I follow your thoughts: What would the nil value mean? Would the key ID be set to nil when the field is left empty?

I'd say that the field that is most similar to `s3AccessKeyId` in the existing code is `repository`, which is also not needed for all repository types, but is a non-optional string. It's just empty when not set.

If `s3AccessKeyId` should be optional in `AppEnvironment`, it will need to be transformed to a non-optional string in `ResticSettings` because `TextField` [takes a binding to a non-optional string](https://developer.apple.com/documentation/swiftui/textfield/init(_:text:)).

sergeymakinen replied:
> Just to make sure I follow your thoughts: What would the nil value mean? Would the key ID be set to nil when the field is left empty?

Yes, nil would mean not being used thus not set at all.

> I'd say that the field that is most similar to `s3AccessKeyId` in the existing code is `repository`, which is also not needed for all repository types, but is a non-optional string. It's just empty when not set.

It's actually never empty since it's mapped to `RESTIC_REPOSITORY` which is required for restic.

> If `s3AccessKeyId` should be optional in `AppEnvironment`, it will need to be transformed to a non-optional string in `ResticSettings` because `TextField` [takes a binding to a non-optional string](https://developer.apple.com/documentation/swiftui/textfield/init(_:text:)).

Right, I think it shouldn't be a problem since controls bind to models and models control environment values, you would only need to slightly adjust a read-write logic inside of `ResticSettings`.

tiwoc replied:
Sounds good 👍 I've pushed an update.

sergeymakinen replied:
Great, thank you!

sergeymakinen reviewed (commented):
I left a minor comment but other than that it's good to go.

sergeymakinen commented:
Thanks @tiwoc! I'm going to make some minor changes in the codebase over a few days and tag a new release.
